### PR TITLE
BUG: IntervalIndex is_non_overlapping_monotonic is incorrect when endpoints are closed on both sides

### LINKED ISSUE #16560: IntervalIndex is_non_overlapping_monotonic is incorrect when endpoints are shared
#### Code Sample, a copy-pastable example if possible

```python
In [2]: idx = pd.IntervalIndex.from_breaks(range(4), closed='both')

In [3]: idx
Out[3]:
IntervalIndex([[0, 1], [1, 2], [2, 3]]
              closed='both',
              dtype='interval[int64]')

In [4]: idx.is_non_overlapping_monotonic
Out[4]: True
```
#### Problem description
The intervals have an overlap in the endpoints, 1 is shared among `[0, 1], [1, 2]` and 2 is shared among `[1, 2], [2, 3]`, but `is_non_overlapping_monotonic` returns `True` indicating that the intervals are non-overlapping.

#### Expected Output
I'd expected `is_non_overlapping_monotonic` to return `False`.  

If my interpretation of "overlapping" is incorrect, and in this context means an interval overlap (i.e. not a point/has positive length), then it'd be nice if this were more explicitly stated.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 34.4.1
Cython: 0.25.2
numpy: 1.13.0rc2
scipy: 0.18.1
xarray: 0.9.5
IPython: 5.3.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: 1.4.4
bottleneck: 1.2.0
tables: 3.4.2
numexpr: 2.6.1
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.0
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.7.3
bs4: None
html5lib: None
sqlalchemy: 1.1.10
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: 0.1.6
pandas_datareader: 0.4.0
</details>

### PR DESCRIPTION
**Pull request by alysivji**
- [x] closes #16560
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -556,8 +556,12 @@ def is_non_overlapping_monotonic(self):
         # must be increasing  (e.g., [0, 1), [1, 2), [2, 3), ... )
         # or decreasing (e.g., [-1, 0), [-2, -1), [-3, -2), ...)
         # we already require left <= right
-        return ((self.right[:-1] <= self.left[1:]).all() or
-                (self.left[:-1] >= self.right[1:]).all())
+        if self.closed == 'both':
+            return ((self.right[:-1] < self.left[1:]).all() or
+                    (self.left[:-1] > self.right[1:]).all())
+        else:
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: don't use the else here

add a comment on these cases (left/right).

neighter case? (short circuit to False)

File: pandas/tests/indexes/test_interval.py [hunk 2]
```diff
@@ -371,10 +371,19 @@ def slice_locs_cases(self, breaks):
...
-        assert index.slice_locs(1, 1) == (0, 2)
+        index = IntervalIndex.from_tuples(
+            [(0, 1), (2, 3), (4, 5)], closed='both')
+        assert index.slice_locs(1, 1) == (0, 1)
         assert index.slice_locs(1, 2) == (0, 2)
 
+    def test_is_non_overlapping_monotonic(self):
+        index = IntervalIndex.from_tuples(
+            [(0, 1), (2, 3), (4, 5)], closed='both')
+        assert index.is_non_overlapping_monotonic == True
```
Comment (jreback) [inline pandas/tests/indexes/test_interval.py]: use is True

File: pandas/tests/indexes/test_interval.py [hunk 3]
```diff
@@ -371,10 +371,19 @@ def slice_locs_cases(self, breaks):
...
+
+        index = IntervalIndex.from_breaks(range(4), closed='both')
+        assert index.is_non_overlapping_monotonic == False
```
Comment (jreback) [inline pandas/tests/indexes/test_interval.py]: use is False

**Review by jreback — CHANGES_REQUESTED** [review]
some changes. add a whatsnew note. 0.21.0 conversion bug fixes

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (alysivji) [inline pandas/core/indexes/interval.py]: Working on this now. Why would the neither case short circuit to False? Can't we have intervals [(0, 1), (1, 2)] where it is non-overlapping and also cases where it is overlapping [(0, 1), (.5, 3)]

Thanks!

Comment (jreback) [inline pandas/core/indexes/interval.py]: this is checking monotonicity as well. so you can do the same for neither.

**Comment by alysivji** [comment]
Fixed the issues from the review, except for changing ```==``` to ```is``` in the asserts. Using ```is``` resulted in assert errors so I kept it the way it was originally.

Added additional tests and updated the whatsnew section in the docs. Thanks.

**Comment by jschendel** [comment]
ping @jreback : This can be closed, as the issue was resolved by #17238

**Comment by gfyoung** [comment]
@jschendel : Confirmed and closing.

**Pull request closed without merging**